apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.SingleChannel.cs

```diff
@@ -821,6 +912,30 @@ void UpdateRetryableIds(IEnumerable<string> allIds, AckError ackError)
                     _ = _retryableIds.TryRemove(item, out _);
                 }
             }
+
+            // This local method is used exclusively for exactly-once delivery flow.
+            // It updates the `_receiptModAckStatusLookup` dictionary with the latest receipt ModAck response status for each Ack ID.
+            void UpdateModAckStatus(IEnumerable<string> allIds, AckError ackError)
+            {
+                var permanentFailureIds = ackError?.PermanentFailureIds ?? Enumerable.Empty<string>();
+                var temporaryFailureIds = ackError?.TemporaryFailureIds ?? Enumerable.Empty<string>();
+                var allFailureIds = permanentFailureIds.Concat(temporaryFailureIds);
+                // Successful Ids = AllIds - (TemporaryFailures + Permanent failures)
+                var successfulIds = allIds.Except(allFailureIds);
+
+                // Update failed status. The value of false represents failure.
+                UpdateReceiptModAckStatus(allFailureIds, false);
+
+                // Update success status. The value of true represents success.
+                UpdateReceiptModAckStatus(successfulIds, true);
+            }
+
+            // This local method is used to determine if the current response is for receipt ModAck. It is used exclusively for exactly-once delivery.
+            // We determine that the response is receipt ModAck, if the following three conditions are met simultaneously:
+            // 1. The response contains at least one Ack ID that is present in the `_receiptModAckStatusLookup` dictionary. This condition helps reduce reduce noise due to extends that may occur simultaneously with receipt ModAck.
+            // 2. The response contains at least one extend ID indicated by the hasExtends flag. Only extends count towards receipt ModAck, not acks or nacks.
+            // 3. The flag _receiptModAckForExactlyOnceDelivery is true. This flag is only set when the lease extension begins.
+            bool IsReceiptModAckResponse(IEnumerable<string> allIds, bool hasExtends) => _receiptModAckStatusLookup.Keys.Any(key => allIds.Contains(key)) && hasExtendIds && _receiptModAckForExactlyOnceDelivery;
```

Comment:
Let's split this line after `=>`

Reply:
I'd also check the flags before the collection, as they'll be much more efficient to check.